File: p328.cpp
#include <bits/stdc++.h>
#include <ListNode.hpp>

using namespace std;

class Solution {
public:
    ListNode* oddEvenList(ListNode* head) {
        if(head == NULL || head->next == NULL)
            return head;
        // the odds
        ListNode* head1 = head;
        // the evens
        ListNode* head2 = head->next;
        
        ListNode* curr1 = head1;
        ListNode* curr2 = head2;
        while(curr1 != NULL && curr1->next != NULL && curr2 != NULL && curr2->next != NULL) {
            curr1->next = curr1->next->next;
            curr2->next = curr2->next->next;
            curr1 = curr1->next;
            curr2 = curr2->next;
        }
        while(curr1 != NULL && curr1->next != NULL) {
            curr1->next = curr1->next->next;
            curr1 = curr1->next;
        }
        while(curr2 != NULL && curr2->next != NULL) {
            curr2->next = curr2->next->next;
            curr2 = curr2->next;
        }
        
        // connect the two linked lists
        ListNode* result = head1;
        ListNode* traverse = result;
        while(traverse->next) {
            traverse = traverse->next;
        }
        traverse->next = head2;
        
        return result;        
    }
};

File: p328.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def oddEvenList(self, head: ListNode) -> ListNode:
        if head is None or head.next is None:
            return head

        odd_head = head
        even_head = head.next

        odd_curr = odd_head
        even_curr = even_head

        while odd_curr and odd_curr.next and even_curr and even_curr.next:
            odd_curr.next = odd_curr.next.next
            even_curr.next = even_curr.next.next
            odd_curr = odd_curr.next
            even_curr = even_curr.next

        while odd_curr and odd_curr.next:
            odd_curr.next = odd_curr.next.next
            if odd_curr.next is None:
                break
            odd_curr = odd_curr.next

        while even_curr and even_curr.next:
            even_curr.next = even_curr.next.next
            even_curr = even_curr.next

        odd_curr.next = even_head

        return odd_head
